src/Features/CSharp/Portable/MakeFieldReadonly/CSharpMakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,42 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.MakeFieldReadonly;
+
+namespace Microsoft.CodeAnalysis.CSharp.MakeFieldReadonly
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class CSharpMakeFieldReadonlyDiagnosticAnalyzer :
+        AbstractMakeFieldReadonlyDiagnosticAnalyzer<IdentifierNameSyntax, ConstructorDeclarationSyntax>
+    {
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeType, SyntaxKind.ClassDeclaration, SyntaxKind.StructDeclaration);
+
+        protected override ISyntaxFactsService GetSyntaxFactsService() =>
+            CSharpSyntaxFactsService.Instance;
+
+        protected override bool IsWrittenTo(IdentifierNameSyntax name, SemanticModel model, CancellationToken cancellationToken)
+            => name.IsWrittenTo();
+
+        protected override bool IsMemberOfThisInstance(SyntaxNode node)
+        {
+            // if it is a qualified name, make sure it is `this.name`
+            if (node.Parent is MemberAccessExpressionSyntax memberAccess)
+            {
+                return memberAccess.Expression is ThisExpressionSyntax;
+            }
+
+            // make sure it isn't in an object initializer
+            if (node.Parent.Parent is InitializerExpressionSyntax)
```

Comment:
what if you have:

```c#
class C {
     int i;

     public C() { new D { X = i };
}
```

(or someting like that).  basically, what happens if it's on the right hand side and in a construct such that .parent.parent is the InitializeExpression?
}

Reply:
➡️ The concern here is *writing* to `C.i` for an instance which is not `this`:

```csharp
class C {
  int i;

  C() {
    // C.i is written to, but this.i is not
    _ = new C { i = 3 };
  }
}
```

As far as I can tell, the situation you are concerned about can only happen if `i` is *read*.